### ISSUE TITLE
Strage behaviour with atomicAdd. Compiler bug (?)

### ISSUE THREAD

**Issue by incardon** [ROCm-Developer-Tools/HIP#2260]
In porting a code to HIP I encounter a strange problem I cannot explain.

I was able to isolate a small code to show the strange behavior I cannot explain.

Unfortunately I have to introduce a couple of elements to show the problem because it look like connected to a combos of stuff.

I have a struct DataBlock that emulate a block of unsigned chars ... [256] and has an operator[]

I have 3 kernels all do the same things (in theory). I will go back to them later. Let see the main function. It is pretty simple I allocate a buffer of 4 data blocks, set it to zero, and create another buffer on host to get back the result. I am launching a kernel with 1 block and 2x2 threads. I transfer the result to host. Finally I have a print function going across the 4 blocks and printing the position of all non zero sets by the kernel.

Now let's see at grid_apply_functor. The kernel has a shared counter set to zero, an atomic add is done after by each thread. This means that on 4 threads each thread (no mater which one) should get one of these numbers 0,1,2,3. I use this number to choose the block and than I set one of the 256 unsigned char in the block. In particular thread (0,0) will set offset 0    (1,0) will set a non zero at 1 (0,1) will set at offset 16 and (1,1) will set at offset 17.

The expected print out should be

DATA2: 0-0      ARRAY: 1
DATA2: 1-1      ARRAY: 1
DATA2: 2-16      ARRAY: 1
DATA2: 3-17      ARRAY: 1

In reality i should see all the numbers 0,1,2,3 in the first column, meaning that all blocks has been set, and 0,1,16,17 in the second the association does not matter.

Although If I run the kernel I obtain:

DATA2: 0-0      ARRAY: 1
DATA2: 0-1      ARRAY: 1
DATA2: 0-16      ARRAY: 1
DATA2: 0-17      ARRAY: 1

That is clearly wrong.

Funny enough if  I now decide to give the buffer as a pointer to unsigned char like in grid_apply_functor3 and I access directly with [] using the formula pos*256+offset. Than I get the correct result

DATA2: 0-0      ARRAY: 1
DATA2: 1-1      ARRAY: 1
DATA2: 2-16      ARRAY: 1
DATA2: 3-17      ARRAY: 1

The situation become even more funny, because there is a way to make the block version working based on what I suppose fooling the compiler to calculate each step. This bring to grid_apply_functor2

I have a bool inactive set to false and a condition to set the flag to true, although the condition is never met and the bool remain false. Than I put the previous code in an if condition always avaluated true in way that probably the compiler cannot know that. Now if I try to set the block inside the if clause I get again the same problem, but if I take the address (that apparently is calculated correctly) and I set it after, in another "if" always evaluated as true (the compiler cannot infer that), than magically the kernel works.
grid_apply_functor2 works producing:

DATA2: 0-0      ARRAY: 1
DATA2: 1-1      ARRAY: 1
DATA2: 2-16      ARRAY: 1
DATA2: 3-17      ARRAY: 1

As far as I have tested you need an atomic add to show the wrong behaviour. If I change 
grid_apply_functor to use ptr_i[offset][offset] it will not show the problem. The same is valid for grid_apply_functor2 as soon as I remove the if condition (x<20 || y < 20) or the last condition (mask != NULL)  or the if(inactive == false) it will show the problem again

The code is tested on a RX Vega 64

```
#include <iostream>
#include <hip/hip_runtime.h>



struct DataBlock
{
    unsigned char block[256];

    __device__ __host__ inline unsigned char &operator[](unsigned int i)
    {
        return block[i];
    }
};


__global__ void grid_apply_functor3(unsigned char * ptr_i)
{
	__shared__ int debug;
	debug = 0;

	int x = threadIdx.x + blockIdx.x * blockDim.x;
	int y = threadIdx.y + blockIdx.y * blockDim.y;

	__syncthreads();

	unsigned int offset = y*16 + x;

	int pos = atomicAdd(&debug,1);
	
	ptr_i[pos*256+offset] = 1;
}

__global__ void grid_apply_functor(DataBlock * ptr_i)
{
	__shared__ int debug;
	debug = 0;

	int x = threadIdx.x + blockIdx.x * blockDim.x;
	int y = threadIdx.y + blockIdx.y * blockDim.y;

	__syncthreads();

	unsigned int offset = y*16 + x;

	int pos = atomicAdd(&debug,1);

	ptr_i[pos][offset] = 1;
}

__global__ void grid_apply_functor2(DataBlock * ptr_i)
{
	__shared__ int debug;
	debug = 0;

	int x = threadIdx.x + blockIdx.x * blockDim.x;
	int y = threadIdx.y + blockIdx.y * blockDim.y;

	bool inactive = false;
		   
	if (x > 20 || y > 20) // Impossible this is evaluated true
	{inactive = true;}

	__syncthreads();

	unsigned char * mask__;

	if (inactive == false) // I always go here
	{
		unsigned int offset = y*16 + x;

		int pos = atomicAdd(&debug,1);

		mask__ = &ptr_i[pos][offset]; // I take the adress here 
	}

	if (mask__ != NULL) // is always different from null
	{
		*mask__ = 1;  // I set here
	}
}


int main(int argc, char * argv[])
{
	float c = 5.0;

	DataBlock * ptr_i;
	hipMalloc(&ptr_i,4*sizeof(DataBlock));

	hipMemset(ptr_i,0,4*sizeof(DataBlock));

	DataBlock * ptr_hi;
	hipHostMalloc(&ptr_hi,4*sizeof(DataBlock));

	dim3 thr;
	dim3 wthr;

	thr.x = 2;
	thr.y = 2;
	thr.z = 1;
	wthr.x = 1;
	wthr.y = 1;
	wthr.z = 1;

	//hipLaunchKernelGGL(HIP_KERNEL_NAME(grid_apply_functor), wthr, thr, 0, 0,ptr_i);                          // Bugged
	hipLaunchKernelGGL(HIP_KERNEL_NAME(grid_apply_functor2), wthr, thr, 0, 0,ptr_i);                           // Not bugged
	//hipLaunchKernelGGL(HIP_KERNEL_NAME(grid_apply_functor3), wthr, thr, 0, 0,(unsigned char *)ptr_i);        // Not bugged

	hipMemcpy(ptr_hi,ptr_i,4*sizeof(DataBlock),hipMemcpyDeviceToHost);

	{
		for (int s = 0 ; s < 4 ; s++)
		{
			for (int k = 0 ; k < 16*16 ; k++)
			{
				if ((int)(int)ptr_hi[s][k] == 1)
				{std::cout << "DATA2: " << s << "-" << k << "     " << " ARRAY: " << (int)ptr_hi[s][k] << std::endl;}
			}
		}
	}


}

```

**Comment by yxsamliu** [comment]
This seems to be a compiler bug about optimization. If you compile with -O0 it will pass. We will take a look. Thanks for reporting.

**Comment by incardon** [comment]
Thanks. Looking forward for the fix

**Comment by incardon** [comment]
Thanks the bug is fixed in 4.5
